form.js

```diff
@@ -1,57 +1,49 @@
 'use strict'
 
+/**
+ * Clamp between two numbers
+ *
+ * @param      {number}  min     The minimum
+ * @param      {number}  max     The maximum
+ * @param      {number}  val     The value to clamp
+ */
+function clamp(min, max, val) {
+  let value = Math.min(max, Math.max(min, val));
+  return value;
+}
+
 /**
  * Fix num cores, allowing blanks to remain
  */
 function fix_num_cores() {
   let node_type_input = $('#batch_connect_session_context_node_type');
-  let node_type = node_type_input.val();
-  let num_cores_input = $('#num_cores');
+  let num_cores_input = $('#batch_connect_session_context_num_cores');
 
   if(num_cores_input.val() === '') {
     return;
   }
 
-  if(node_type === 'hugemem') {
-    set_ppn_owens_hugemem(num_cores_input);
-  } else {
-    set_ppn_owens_regular(num_cores_input);
-  }
+  set_ppn_by_node_type(node_type_input, num_cores_input);
 }
 
-/**
- * Sets the PPN limits available for Owens hugemem nodes.
- *
- * hugemem reservations are always assigned the full node
- *
- * @param      {element}  num_cores_input  The input for num_cores
- */
-function set_ppn_owens_hugemem(num_cores_input) {
-  const NUM_CORES = 48;
-  num_cores_input.attr('max', NUM_CORES);
-  num_cores_input.attr('min', NUM_CORES);
-  num_cores_input.val(NUM_CORES);
-}
+function set_ppn_by_node_type(node_type_input, num_cores_input){
+  let data = node_type_input.find(':selected').data();
 
-/**
- * Sets the PPN limits available for non hugemem Owens nodes.
- *
- * @param      {element}  num_cores_input  The input for num_cores
- */
-function set_ppn_owens_regular(num_cores_input) {
-  const NUM_CORES = 28;
-  num_cores_input.attr('max', NUM_CORES);
-  num_cores_input.attr('min', 0);
-  num_cores_input.val(Math.min(NUM_CORES, num_cores_input.val()));
-}
+  num_cores_input.attr('max', data.maxPpn);
+  num_cores_input.attr('min', data.minPpn);
 
+  // Clamp value between min and max
+  num_cores_input.val(
+    clamp(data.minPpn, data.maxPpn, num_cores_input.val())
```

Comment:
I understand the desire here. The input is set to 40 for pitzer, then you change a dropdown and want to restrict to owens, but the max ppn is 28, so you change the value to 28. The problem is with this logic it seems that if you then switch back to pitzer, the value remains 28. So previously what a user had done was "choose all the cores for a node" i.e. "give me the whole node" and now they have Pitzer selected with part of the node. So I'm not sure if that is ideal. Though I think that in the past in other apps we had discussed if you leave it blank, you get the whole node. Is that now if you keep it to 0?

Reply:
I'm noticing a bug where the maxPpn doesn't change immediately when you change the cluster. Is that the problem here? I realize I never call `fix_num_cores` in the cluster change callback.
It looks like the default num_cores we have for the current rstudio is 1. It seems like the only way to make it blank and get the full node is to manually delete the value in the number_field. Maybe we can make 0 mean the full node? Or I can default the value to maxPpn?